b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+from ..scan.folder import B2Folder
+from ..scan.path import B2Path
+from ..scan.policies import DEFAULT_SCAN_MANAGER, ScanPoliciesManager
+from ..scan.report import Report
+from ..scan.scan import zip_folders
+from .setting import ReplicationRule
+from .types import ReplicationStatus
+
+
+@dataclass(frozen=True)
+class AbstractReplicationAttrs(metaclass=ABCMeta):
+    """
+    Some attributes of files which are meaningful
+    for replication monitoring and troubleshooting.
+    """
+
+    @abstractclassmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'AbstractReplicationAttrs':
+        pass
+
+
+@dataclass(frozen=True)
+class ReplicationAttrs(AbstractReplicationAttrs):
+    """
+    Some attributes of source and destination files and their relations
+    which are meaningful for replication monitoring and troubleshooting.
+    """
+
+    # source attrs
+    source_replication_status: Optional[ReplicationStatus] = None
+    source_has_hide_marker: Optional[bool] = None
+    source_has_sse_c_enabled: Optional[bool] = None
+    source_has_large_metadata: Optional[bool] = None
+    source_has_file_retention: Optional[bool] = None
+    source_has_legal_hold: Optional[bool] = None
+
+    # destination attrs
+    destination_replication_status: Optional[ReplicationStatus] = None
+
+    # source & destination relation attrs
+    metadata_differs: Optional[bool] = None
+
+    LARGE_METADATA_SIZE: ClassVar[int] = 2048
+
+    @classmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'ReplicationAttrs':
+        params = {}
+
+        if source_file:
+            source_file_version = source_file.selected_version
+            params.update(
+                {
+                    'source_replication_status':
+                        source_file_version.replication_status,
+                    'source_has_hide_marker':
+                        source_file.is_visible(),
+                    'source_has_sse_c_enabled':
+                        source_file_version.server_side_encryption.mode == EncryptionMode.SSE_C,
+                    'source_has_large_metadata':
+                        source_file_version.headers_size >= cls.LARGE_METADATA_SIZE,
+                    'source_has_file_retention':
+                        source_file_version.file_retention is not NO_RETENTION_FILE_SETTING,
+                    'source_has_legal_hold':
+                        source_file_version.legal_hold is not LegalHold.UNSET,
+                }
+            )
+
+        if destination_file:
+            params.update(
+                {
+                    'destination_replication_status':
+                        destination_file.selected_version.replication_status,
+                }
+            )
+
+        if source_file and destination_file:
+            params.update(
+                {
+                    'metadata_differs':
```

Comment:
I think the essence of that one is a file that has the same size but different metadata, however here you only check for metadata and not for the hash.

I'll ask around, see how it actually works.

Reply:
yes, you should compare hash too

Reply:
https://github.com/reef-technologies/b2-sdk-python/commit/057a68d9fe4982017fae2def61ff13bc64dff3d1 ?